### PR TITLE
chore: preserve assessed fees even on insufficient token balances

### LINKED ISSUE #11566: Replay events from round 162559605 and validate record files for differential testing - 15 min
Replay events from round 162559605 and validate record files for differential testing - 15 min
issues found
- Mismatched field counts  (expected but not find [tokenTransferLists, assessed_custom_fees]) between expected receipt {
- Mismatched values, expected '2066566', got '2066567' - Matching field 'amount' Matching field 'transferList'  ==> expected: <2066566> but was: <2066567>
- Mismatched values, expected '<ByteString@2a037324 size=32 contents="`\025\016u\035\216\n\263\255h\211\305\254J4\001\017[\223J\312Q^\373\347\311\2771\373\225.\342">', got '<ByteString@69eb86b4 size=32 contents="C\343\265/~\207 \312\301\244\207\311&~\033G\213\026\351;/\330\022\224~y\v%\222v\313\372">' - Matching field 'contractCallResult' Matching field 'contractCallResult'  ==> expected: <<ByteString@2a037324 size=32 contents="`\025\016u\035\216\n\263\255h\211\305\254J4\001\017[\223J\312Q^\373\347\311\2771\373\225.\342">> but was: <<ByteString@69eb86b4 size=32 contents="C\343\265/~\207 \312\301\244\207\311&~\033G\213\026\351;/\330\022\224~y\v%\222v\313\372">>
- Mismatched values, expected '<ByteString@3b0ca5e1 size=32 contents="_\"_\212\016\377\352\005B\002\312\331\261u\023N\373\022\b\375*?CmL\313\321\016Cw\367\245">', got '<ByteString@5bb3131b size=32 contents="\376\373\305\247z\216o?T\2557@j\\\f\352\322\004~\217\025:a\202}a\000\325Fla\206">' - Matching field 'contractCallResult' Matching field 'contractCallResult'  ==> expected: <<ByteString@3b0ca5e1 size=32 contents="_\"_\212\016\377\352\005B\002\312\331\261u\023N\373\022\b\375*?CmL\313\321\016Cw\367\245">> but was: <<ByteString@5bb3131b size=32 contents="\376\373\305\247z\216o?T\2557@j\\\f\352\322\004~\217\025:a\202}a\000\325Fla\206">>
- Mismatched values, expected '<ByteString@ba1f559 size=32 contents="\352B=\222\221\326\245\231\346_\3473\216)\t!\321\002\233l\312\337\330\322\244\027\370\211@w\343\353">', got '<ByteString@3f4f9acd size=32 contents="\233\267\036)(\246O\220\302/\321\324\247\223\006\003X\333\265@\022\217f\320\332\254 \006\325\332i3">' - Matching field 'contractCallResult' Matching field 'contractCallResult'  ==> expected: <<ByteString@ba1f559 size=32 contents="\352B=\222\221\326\245\231\346_\3473\216)\t!\321\002\233l\312\337\330\322\244\027\370\211@w\343\353">> but was: <<ByteString@3f4f9acd size=32 contents="\233\267\036)(\246O\220\302/\321\324\247\223\006\003X\333\265@\022\217f\320\332\254 \006\325\332i3">>
- Transaction fees '2407392' and '2407348' varied by more than 1 tinybar - Matching field 'transactionFee'  ==> expected: <true> but was: <false>

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #11566

### CODE REVIEW

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AdjustFungibleTokenChangesStep.java [hunk 1]
```diff
@@ -164,21 +172,45 @@ private void modifyAggregatedAllowances(
...
+            try {
+                adjustBalance(rel, account, amount, tokenRelStore, accountStore);
+            } catch (HandleException e) {
+                if (e.getStatus() == INSUFFICIENT_TOKEN_BALANCE
+                        && effectivePaymentWasMade(rel.accountIdOrThrow(), rel.tokenIdOrThrow(), assessedCustomFees)) {
+                    throw new HandleException(INSUFFICIENT_SENDER_ACCOUNT_BALANCE_FOR_CUSTOM_FEE);
+                }
+                throw e;
+            }
+        }
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AdjustFungibleTokenChangesStep.java:201]: Javadoc here might be helpful 🙂

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/TransferContext.java [hunk 2]
```diff
@@ -88,4 +89,8 @@ public interface TransferContext {
      * @param assessedCustomFee the assessed custom fee
      */
     void addToAssessedCustomFee(AssessedCustomFee assessedCustomFee);
+
+    List<AssessedCustomFee> getAssessedCustomFees();
+
+    boolean isEnforceMonoServiceRestrictionsOnAutoCreationCustomFeePayments();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/TransferContext.java:95]: Same here.

**Review by mhess-swl — DISMISSED** [review]

**Review by iwsimon — APPROVED** [review]
LGTM

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/TransferContext.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/TransferContext.java:111]: Done.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AdjustFungibleTokenChangesStep.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AdjustFungibleTokenChangesStep.java]: Done.

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! tyvm @tinker-michaelj

**Pull request merged**